### PR TITLE
Lesson Group: Update script edit gui

### PR DESCRIPTION
**Pull request by dmcavoy**
Makes the Script Edit Gui work with lesson groups!

I renamed anything that was Flex Category related to Lesson Group.

I changed the way the data was getting sent into the editor which led to many of the changes you will see especially the ones in editorRedux. We used to have lessons information and then separate flex_category_map. I didn't feel like it made sense for those to be separate anymore with lesson groups so I replaced the lessons information with lessonGroups which contain lessons and got rid of the separate flex_category_map.

Since levelbuilders can now add their own new lesson groups (which they could not do before with flex categories) I changed the way you add a lesson group to allow for this. 

If you click the Add Lesson Group button:
Screen Shot 2020-05-15 at 4 51 25 PM

You will see this:
Screen Shot 2020-05-15 at 4 51 33 PM

## Links

- [jira](https://codedotorg.atlassian.net/browse/PLAT-130)

## Testing story

Updated existing tests to match changes.
Added tests for NewLessonGroupInput and lesson_group summarize method.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/scripts_controller.rb [hunk 1]
```diff
@@ -112,11 +112,10 @@ def edit
       beta: beta,
       betaWarning: beta_warning,
       levelKeyList: beta && Level.key_list,
-      stageLevelData: @script_file,
+      lessonLevelData: @script_file,
```
Comment (dmcavoy) [inline dashboard/app/controllers/scripts_controller.rb:115]: Is it ok to make this change over or should i keep stageLevelData for a little bit?

Comment (davidsbailey) [reply dashboard/app/controllers/scripts_controller.rb:115]: this is fine, because no one is currently relying on this UI.

File: dashboard/app/models/lesson_group.rb [hunk 2]
```diff
@@ -30,4 +30,27 @@ class LessonGroup < ApplicationRecord
...
+
+    summary
+  end
+
+  def summarize_for_edit
+    include_lessons = false
+    summary = summarize(include_lessons)
+    summary[:lessons] = lessons.map(&:summarize_for_edit)
+    summary
+  end
```
Comment (davidsbailey) [inline dashboard/app/models/lesson_group.rb:58]: these two methods seem a bit more complex than they need to be. It looks like the code which is used to summarize lessons for displaying to the user bypasses the lesson group and looks for the lessons directly: https://github.com/code-dot-org/code-dot-org/blob/73fe9bc5466cbc8b38fcebb3fc47cd74ed9e1382/dashboard/app/models/script.rb#L1443-L1445

If there are no other planned uses for lesson_group.summarize, then `summarize_for_edit` would be the only caller, in which case here is what I'd recommend:
1. include_bonus_levels appears to be unused and looks like it can be removed. its default value in lesson.summarize is already `false`, and I don't see any place we set it to true.
2. don't bother filtering out lessons with a visible_after date in the future, unless there's a specific reason you want to hide these in the gui editor
3. `user` appears to always be `nil`. You may be able to remove this param, if you haven't already as part of (2).

File: apps/src/lib/script-editor/editorRedux.js [hunk 3]
```diff
@@ -18,140 +18,164 @@ const MOVE_LESSON = 'scriptEditor/MOVE_LESSON';
...
+  position,
+  groupKey,
   groupName
 });
 
-export const addLesson = (position, lessonName) => ({
+export const addLesson = (group, lessonName) => ({
   type: ADD_LESSON,
-  position,
+  group,
```
Comment (davidsbailey) [inline apps/src/lib/script-editor/editorRedux.js:40]: it's a bit hard to follow what position and group mean here. in practice, are these redundant, meaning that the first lesson group always has a `group` of `0` and a `position` of `1`? this would be helpful to note in the comments, perhaps on this method since it uses the `position` param.

If you want you could rename these to groupPosition and groupIndex, but that could be a lot of renaming so it could be fine to just explain in comments.

File: apps/src/lib/script-editor/editorRedux.js [hunk 4]
```diff
@@ -160,50 +184,67 @@ function updateLevelPositions(levels) {
...
+      const level = levels.splice(action.level - 1, 1)[0];
       updateLevelPositions(levels);
-      const newLevels = newState[action.newLesson - 1].levels;
+
+      // add level to new lesson
+      let newGroupPosition = null;
+      newState.forEach(lessonGroup => {
+        lessonGroup.lessons.forEach(lesson => {
+          if (lesson.position === action.newLesson) {
+            newGroupPosition = lessonGroup.position;
```
Comment (davidsbailey) [inline apps/src/lib/script-editor/editorRedux.js:215]: this was surprising to read -- newLesson is a position, not an index? or does position === index? this looks like it may have been this way before you touched it, but maybe a comment on `moveLevelToLesson` would help explain what `action.newLesson` represents, now that you've already taken the time to understand it.

File: apps/src/lib/script-editor/editorRedux.js [hunk 5]
```diff
@@ -213,112 +254,134 @@ function lessons(state = [], action) {
...
-        newState[index].flex_category = newState[swap].flex_category;
+        // add to new lesson group
+        const groupSwapIndex =
+          action.direction === 'up' ? groupIndex - 1 : groupIndex + 1;
+
+        // Remove the lesson
+        const oldLessons = newState[groupIndex].lessons;
+        const curLesson = oldLessons.splice(lessonIndex, 1)[0];
+
+        // add lesson
```
Comment (davidsbailey) [inline apps/src/lib/script-editor/editorRedux.js]: it would be helpful to clean up these comments a bit. the `// add to new lesson group` looks out of place. the other two could read better as something like:
```
// Remove the lesson from the old lesson group
...
// Add the lesson to the new lesson group
```

File: apps/src/lib/script-editor/editorRedux.js [hunk 6]
```diff
@@ -213,112 +254,134 @@ function lessons(state = [], action) {
...
-
+    case SET_LESSON_GROUP: {
+      // Remove the lesson
+      const oldLessons = newState[action.oldGroup - 1].lessons;
+      const curLesson = oldLessons.splice(
+        action.lesson - oldLessons[0].position,
+        1
+      )[0];
+
+      // add lesson
```
Comment (davidsbailey) [inline apps/src/lib/script-editor/editorRedux.js]: could improve comments here similarly to above

File: apps/src/lib/script-editor/LessonGroup.jsx [hunk 7]
```diff
@@ -229,31 +231,27 @@ class FlexGroup extends Component {
...
-          <div style={styles.flexCategorySelector}>
-            <FlexCategorySelector
-              labelText="New Flex Category"
-              confirmButtonText="Create"
-              onConfirm={this.createFlexCategory}
-              onCancel={this.hideFlexCategorySelector}
-            />
-          </div>
+        {this.state.addingLessonGroup && (
+          <NewLessonGroupInput
```
Comment (davidsbailey) [inline apps/src/lib/script-editor/LessonGroup.jsx:246]: nice extraction

File: apps/src/lib/script-editor/LessonGroup.jsx [hunk 8]
```diff
@@ -41,89 +40,97 @@ const styles = {
...
-    height: 30,
-    marginBottom: 30
   }
 };
 
 // Replace ' with \'
 const escape = str => str.replace(/'/, "\\'");
 
-class FlexGroup extends Component {
+class LessonGroup extends Component {
```
Comment (davidsbailey) [inline apps/src/lib/script-editor/LessonGroup.jsx:49]: this is really a list of lesson groups, right? can you please note this in the comments since the name is somewhat misleading? Renaming the component to LessonGroups or something else less misleading would also be a more ambitious option.

File: apps/src/lib/script-editor/LessonCard.jsx [hunk 9]
```diff
@@ -235,6 +246,7 @@ export class UnconnectedLessonCard extends Component {
           <OrderControls
             type={ControlTypes.Lesson}
             position={lesson.position}
+            parentPosition={lessonGroupPosition}
```
Comment (davidsbailey) [inline apps/src/lib/script-editor/LessonCard.jsx:249]: for another time... it's a bit weird for the OrderControls to need to know the parentPosition. however, I can't blame you because it OrderControls doesn't really need to know its own position either, and you are following the same pattern. Ideally LessonCard would define:
```js
  move = direction => {
    this.props.moveLesson(this.props.parentPosition, this.props.position, direction)
  }
  ...
  render() {
    <OrderControls move = this.props.moveLesson/>
  }
  ...
  connect(
    dispatch => ({
      moveLesson(position, parentPosition, direction) {
        dispatch(moveLesson(parentPosition, position, direction));
      }
    })
  )
```
this would let you skip all of the type-checking logic in in the `connect` method on OrderControls too (explicit type checking is a good sign that things may not be encapsulated properly).

**Review by davidsbailey — COMMENTED** [review]
Awesome work @dmcavoy ! Getting flex categories working here was a huge pain, so I can only imagine that undoing that and redoing it a totally different way was at least as cumbersome. 

I've got a few comments inline and have tried to indicate that some of them can be treated as optional. If there are any changes (renames ore refactorings) that you want to make as a follow-up, feel free to spin off a new PR so that this one doesn't get too much bigger.

Since I'm not quite sure what the summarize code will end up looking like, I'd like to review again once you've had a chance to address comments there.

File: apps/src/lib/script-editor/LessonGroup.jsx (hunk 8 above)
Comment (dmcavoy) [reply apps/src/lib/script-editor/LessonGroup.jsx:49]: Made this change in a follow up PR: https://github.com/code-dot-org/code-dot-org/pull/34875

File: apps/src/lib/script-editor/editorRedux.js (hunk 4 above)
Comment (dmcavoy) [reply apps/src/lib/script-editor/editorRedux.js:215]: Made this change in a follow up PR: https://github.com/code-dot-org/code-dot-org/pull/34875

File: apps/src/lib/script-editor/editorRedux.js (hunk 3 above)
Comment (dmcavoy) [reply apps/src/lib/script-editor/editorRedux.js:40]: Made this change in a follow up PR: https://github.com/code-dot-org/code-dot-org/pull/34875

File: dashboard/app/models/lesson_group.rb (hunk 2 above)
Comment (dmcavoy) [reply dashboard/app/models/lesson_group.rb:58]: I added a comment (as we had discussed) about why summarize is there even though it not really used right now.

File: apps/src/lib/script-editor/LessonCard.jsx (hunk 9 above)
Comment (dmcavoy) [reply apps/src/lib/script-editor/LessonCard.jsx:249]: Working on this in a follow up PR: https://github.com/code-dot-org/code-dot-org/pull/34893 (it grew bigger than I wanted to tackle in this PR)

File: dashboard/app/models/lesson_group.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/app/models/lesson_group.rb:58]: new comments look good!

File: dashboard/test/models/lesson_group_test.rb [hunk 10]
```diff
@@ -14,4 +14,39 @@ class LessonGroupTest < ActiveSupport::TestCase
...
+  test 'summarize should exclude lessons if include_lessons is false' do
+    script = create :script
+    lesson_group = create :lesson_group, key: 'my-lesson-group', user_facing: true, position: 1, script: script
+    lesson = create :lesson, name: "Lesson1", script: script, lesson_group: lesson_group, absolute_position: 1
+    create(:script_level, script: script, lesson: lesson)
+
+    assert_nil lesson_group.summarize(false)[:lessons]
+  end
+
+  test 'should summarize lesson groups for edit' do
```
Comment (davidsbailey) [inline dashboard/test/models/lesson_group_test.rb:39]: 👍

**Review by davidsbailey — APPROVED** [review]
Awesome work @dmcavoy ! looking forward to reviewing the follow-up PRs.
